From 9cf6656bfc059af94c57957db356785ffa18dceb Mon Sep 17 00:00:00 2001
From: Claudio DeSouza <cdesouza@chromium.org>
Date: Tue, 22 Oct 2024 18:07:53 +0000
Subject: [PATCH] `CloseBubbleOnTabActivationHelper` double deregistering

This came with ubsan recently, where `OnTabStripModelDestroyed` get
called for `CloseBubbleOnTabActivationHelper`, and it tries to
deregister using the member `tab_strip_model()`, and in some cases,
`Browser::tab_strip_model()` is already gone.

This CL undoes the local `RemoveObserver` call, that has been added
by mistake to `CloseBubbleOnTabActivationHelper`, as this call is
supposed to be handled by the base class. Furthermore, this CL reverts
https://crrev.com/c/3753347, as that particular CL was addressing UaF
failures caused by the introduction of `RemoveObserver` to this class.

Bug: 40248746
Change-Id: Idc474fea4066d8d961e732aeabd53829ce65e420
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5924957
Reviewed-by: Peter Kasting <pkasting@chromium.org>
Reviewed-by: Claudio DeSouza <cdesouza@chromium.org>
Commit-Queue: Claudio DeSouza <cdesouza@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1372178}
---

--- a/chrome/browser/ui/views/bubble/webui_bubble_manager.cc
+++ b/chrome/browser/ui/views/bubble/webui_bubble_manager.cc
@@ -6,6 +6,7 @@
 
 #include "base/notimplemented.h"
 #include "base/timer/timer.h"
+#include "chrome/browser/ui/browser.h"
 #include "chrome/browser/ui/browser_list.h"
 #include "chrome/browser/ui/webui/top_chrome/webui_contents_warmup_level_recorder.h"
 #include "chrome/browser/ui/webui/top_chrome/webui_url_utils.h"
@@ -69,7 +70,8 @@
   if ((!disable_close_bubble_helper_) &&
       BrowserList::GetInstance()->GetLastActive()) {
     close_bubble_helper_ = std::make_unique<CloseBubbleOnTabActivationHelper>(
-        bubble_view_.get(), BrowserList::GetInstance()->GetLastActive());
+        bubble_view_.get(),
+        BrowserList::GetInstance()->GetLastActive()->tab_strip_model());
   }
 
   if (identifier)
--- a/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.cc
+++ b/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.cc
@@ -10,17 +10,16 @@
 
 CloseBubbleOnTabActivationHelper::CloseBubbleOnTabActivationHelper(
     views::BubbleDialogDelegateView* owner_bubble,
-    Browser* browser)
-    : owner_bubble_(owner_bubble), browser_(browser) {
-  DCHECK(owner_bubble_);
-  DCHECK(browser_);
-  browser_->tab_strip_model()->AddObserver(this);
+    TabStripModel* tab_strip_model)
+    : owner_bubble_(owner_bubble) {
+  CHECK(owner_bubble_);
+  CHECK(tab_strip_model);
+  // `AddObserver` called asymmetrically, with no `RemoveObserver` call in this
+  // class as `TabStripModel` removes itself.
+  tab_strip_model->AddObserver(this);
 }
 
-CloseBubbleOnTabActivationHelper::~CloseBubbleOnTabActivationHelper() {
-  if (browser_)
-    browser_->tab_strip_model()->RemoveObserver(this);
-}
+CloseBubbleOnTabActivationHelper::~CloseBubbleOnTabActivationHelper() = default;
 
 void CloseBubbleOnTabActivationHelper::OnTabStripModelChanged(
     TabStripModel* tab_strip_model,
@@ -36,10 +35,3 @@
     owner_bubble_ = nullptr;
   }
 }
-
-void CloseBubbleOnTabActivationHelper::OnTabStripModelDestroyed(
-    TabStripModel* tab_strip_model) {
-  DCHECK(browser_);
-  browser_->tab_strip_model()->RemoveObserver(this);
-  browser_ = nullptr;
-}
--- a/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.h
+++ b/chrome/browser/ui/views/close_bubble_on_tab_activation_helper.h
@@ -8,8 +8,6 @@
 #include "base/memory/raw_ptr.h"
 #include "chrome/browser/ui/tabs/tab_strip_model_observer.h"
 
-class Browser;
-
 namespace views {
 class BubbleDialogDelegateView;
 }
@@ -19,11 +17,12 @@
 // add, close or change the active tab.
 class CloseBubbleOnTabActivationHelper : public TabStripModelObserver {
  public:
-  // It is the expectation of this class that |bubble| and |browser| should
-  // outlive it. The recommended usage is for |bubble| to own |this|.
+  // It is the expectation of this class that `owner_bubble` and
+  // `tab_strip_model` should outlive it. The recommended usage is for `bubble`
+  // to own `this`.
   CloseBubbleOnTabActivationHelper(
       views::BubbleDialogDelegateView* owner_bubble,
-      Browser* browser);
+      TabStripModel* tab_strip_model);
 
   CloseBubbleOnTabActivationHelper(const CloseBubbleOnTabActivationHelper&) =
       delete;
@@ -37,11 +36,9 @@
       TabStripModel* tab_strip_model,
       const TabStripModelChange& change,
       const TabStripSelectionChange& selection) override;
-  void OnTabStripModelDestroyed(TabStripModel* tab_strip_model) override;
 
  private:
   raw_ptr<views::BubbleDialogDelegateView> owner_bubble_;  // weak, owns me.
-  raw_ptr<Browser> browser_;
 };
 
 #endif  // CHROME_BROWSER_UI_VIEWS_CLOSE_BUBBLE_ON_TAB_ACTIVATION_HELPER_H_
--- a/chrome/browser/ui/views/profiles/profile_menu_view_base.cc
+++ b/chrome/browser/ui/views/profiles/profile_menu_view_base.cc
@@ -530,7 +530,7 @@
     : BubbleDialogDelegateView(anchor_button, views::BubbleBorder::TOP_RIGHT),
       browser_(browser),
       anchor_button_(anchor_button),
-      close_bubble_helper_(this, browser) {
+      close_bubble_helper_(this, browser->tab_strip_model()) {
   SetButtons(static_cast<int>(ui::mojom::DialogButton::kNone));
   // TODO(tluk): Remove when fixing https://crbug.com/822075
   // The sign in webview will be clipped on the bottom corners without these
